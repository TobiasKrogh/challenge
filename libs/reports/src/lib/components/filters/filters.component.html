<ng-container [formGroup]="form">
  <select class="form-select" formControlName="projectId">
    <option value="">
      {{ 'REPORTS.ALL_PROJECTS' | translate }}
    </option>
    <option *ngFor="let project of projectsAsArray" [value]="project.projectId">
      {{ project.name }}
    </option>
  </select>

  <select class="form-select" formControlName="gatewayId">
    <option value="">
      {{ 'REPORTS.ALL_GATEWAYS' | translate }}
    </option>
    <option *ngFor="let gateway of gatewaysAsArray" [value]="gateway.gatewayId">
      {{ gateway.name }}
    </option>
  </select>

  <input
    type="hidden"
    formControlName="from"
    ngbDatepicker
    #fromCalendar="ngbDatepicker"
    container="body"
    placement="bottom-right"
    [positionTarget]="fromDateButton"
    [footerTemplate]="fromFooter"
    [minDate]="minDate"
    [maxDate]="maxDate"
    [startDate]="minDate"
  />
  <button
    #fromDateButton
    class="date-picker"
    (click)="fromCalendar.toggle()"
    type="button"
  >
    {{ form.from ? (form.from | formatDate) : 'From Date' }}
  </button>

  <input
    type="hidden"
    formControlName="to"
    ngbDatepicker
    #toCalendar="ngbDatepicker"
    container="body"
    placement="bottom-right"
    [positionTarget]="toDateButton"
    [footerTemplate]="toFooter"
    [minDate]="form.from ?? minDate"
    [maxDate]="maxDate"
  />
  <button
    #toDateButton
    class="date-picker"
    (click)="toCalendar.toggle()"
    type="button"
  >
    {{ form.to ? (form.to | formatDate) : 'To Date' }}
  </button>
</ng-container>

<button type="button" (click)="report()">
  {{ 'REPORTS.GENERATE' | translate }}
</button>

<ng-template #fromFooter>
  <div class="date-reset">
    <button (click)="form.resetControl('from'); fromCalendar.close()">
      {{ 'RESET' | translate }}
    </button>
  </div>
</ng-template>

<ng-template #toFooter>
  <div class="date-reset">
    <button (click)="form.resetControl('to'); toCalendar.close()">
      {{ 'RESET' | translate }}
    </button>
  </div>
</ng-template>
