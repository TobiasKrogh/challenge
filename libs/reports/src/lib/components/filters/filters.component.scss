:host {
  display: flex;
  height: 2rem;
  gap: 1.5rem;
}

select {
  background-color: var(--hex-color-cyan);
  background-image: url(/assets/images/icons/arrow-down.v1.svg);
  border-radius: var(--border-radius--default);
  color: var(--hex-color-white);
  font: var(--font-chart-filter-action);
  padding-top: 0;
  padding-bottom: 0;
  position: relative;
  width: 8rem;

  &:focus {
    box-shadow: none;
  }
}

.date {
  &-picker {
    background-color: var(--hex-color-cyan);
    background-image: url(/assets/images/icons/calendar.v1.svg);
    background-repeat: no-repeat;
    background-position: right 0.75rem center;
    text-align: left;
  }

  &-reset {
    display: flex;
    justify-content: flex-end;
    margin: 0.5rem;
  }
}

button {
  background-color: var(--hex-color-blue--medium);
  border-radius: var(--border-radius--default);
  color: var(--hex-color-white);
  font: var(--font-chart-filter-action);
  height: 2rem;
  padding: 0 0.625rem;
  width: 8rem;
}
