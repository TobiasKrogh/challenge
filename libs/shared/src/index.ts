export * from './lib/components/message/message.component';
export * from './lib/functions/normalize-date';
export * from './lib/pipes/format-amount.pipe';
export * from './lib/pipes/format-date.pipe';
