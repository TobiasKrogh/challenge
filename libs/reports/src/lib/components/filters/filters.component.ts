import {
  ChangeDetectionStrategy,
  Component,
  EventEmitter,
  Input,
  Output,
} from '@angular/core';
import { NgbDateStruct } from '@ng-bootstrap/ng-bootstrap';
import {
  IFilters,
  IGateway,
  IProject,
  TIndexedGateways,
  TIndexedProjects,
} from '@challenge/api';
import { FiltersForm } from './filters.form';

@Component({
  selector: 'reports-filters',
  templateUrl: './filters.component.html',
  styleUrls: ['./filters.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FiltersComponent {
  @Output() filtersChange = new EventEmitter<IFilters>();

  @Input() gateways: TIndexedGateways | null = {};
  @Input() projects: TIndexedProjects | null = {};

  public minDate: NgbDateStruct = {
    year: 2021,
    month: 1,
    day: 1,
  };

  public maxDate: NgbDateStruct = {
    year: 2021,
    month: 12,
    day: 31,
  };

  public form = new FiltersForm();

  public generate(): void {
    this.filtersChange.emit(this.form.normalized);
  }

  public get gatewaysAsArray(): IGateway[] {
    return Object.values(this.gateways ?? {});
  }

  public get projectsAsArray(): IProject[] {
    return Object.values(this.projects ?? {});
  }
}
