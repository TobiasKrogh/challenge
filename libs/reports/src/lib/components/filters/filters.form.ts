import { FormControl, FormGroup } from '@angular/forms';
import { NgbDateStruct } from '@ng-bootstrap/ng-bootstrap';
import { IFilters } from '@challenge/api';
import { normalizeDate } from '@challenge/shared';

export class FiltersForm extends FormGroup {
  constructor() {
    super({
      gatewayId: new FormControl(''),
      projectId: new FormControl(''),
      from: new FormControl(''),
      to: new FormControl(''),
    });
  }

  public get normalized(): IFilters {
    return Object.fromEntries(
      new Map(
        Object.entries(this.controls)
          .filter(([, { value }]) => !!value)
          .map(([key, control]) => [key, normalizeDate(control.value)])
      )
    );
  }

  public get from(): NgbDateStruct | undefined {
    return this.value.from ?? undefined;
  }

  public get to(): NgbDateStruct | undefined {
    return this.value.to ?? undefined;
  }

  public resetControl(target: string): void {
    return this.patchValue({ [target]: null });
  }
}
