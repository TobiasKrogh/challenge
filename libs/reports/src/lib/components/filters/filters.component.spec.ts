import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { NgbDatepickerModule } from '@ng-bootstrap/ng-bootstrap';
import { TranslatePipe } from '@ngx-translate/core';
import { FormatDatePipe } from '@challenge/shared';
import { MockModule, MockPipes } from 'ng-mocks';
import { MOCK_GATEWAY, MOCK_PROJECT } from '../../../testing/data.mocks';

import { FiltersComponent } from './filters.component';

describe('FiltersComponent', () => {
  let component: FiltersComponent;
  let fixture: ComponentFixture<FiltersComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ReactiveFormsModule, MockModule(NgbDatepickerModule)],
      declarations: [
        MockPipes(TranslatePipe),
        FormatDatePipe,
        FiltersComponent,
      ],
    }).compileComponents();

    fixture = TestBed.createComponent(FiltersComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  describe('When rendering', () => {
    it('Then it matches snapshot', () => {
      component.gateways = {
        [MOCK_GATEWAY.gatewayId]: MOCK_GATEWAY,
      };
      component.projects = {
        [MOCK_PROJECT.projectId]: MOCK_PROJECT,
      };
      fixture.detectChanges();
      expect(fixture.debugElement).toMatchSnapshot();
    });

    it('Then it matches snapshot (empty data lists)', () => {
      component.gateways = null;
      component.projects = null;
      fixture.detectChanges();
      expect(fixture.debugElement).toMatchSnapshot();
    });
  });

  describe('When generating report', () => {
    let normalizedGetSpy: jest.SpyInstance;

    beforeEach(() => {
      jest.spyOn(component.filtersChange, 'emit');
      normalizedGetSpy = jest.spyOn(component.form, 'normalized', 'get');
      normalizedGetSpy.mockReturnValue({});
      component.generate();
    });

    it('Then the event emitter emits properly', () => {
      expect(component.filtersChange.emit).toHaveBeenCalledWith({});
    });

    it('Then the normalied form data is retrieved', () => {
      expect(normalizedGetSpy).toHaveBeenCalledWith();
    });
  });
});
