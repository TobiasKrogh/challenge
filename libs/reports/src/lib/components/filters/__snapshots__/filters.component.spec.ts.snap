// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FiltersComponent When rendering Then it matches snapshot (empty data lists) 1`] = `
DebugElement {
  "nativeNode": <div
    id="root2"
  >
    <form
      class="ng-untouched ng-pristine ng-valid"
      novalidate=""
    >
      <select
        class="form-select ng-untouched ng-pristine ng-valid"
        formcontrolname="projectId"
      >
        <option
          value=""
        >
            
        </option>
        
      </select>
      <select
        class="form-select ng-untouched ng-pristine ng-valid"
        formcontrolname="gatewayId"
      >
        <option
          value=""
        >
            
        </option>
        
      </select>
      <input
        class="ng-untouched ng-pristine ng-valid"
        container="body"
        formcontrolname="from"
        ngbdatepicker=""
        placement="bottom-right"
        type="hidden"
      />
      
      <button
        class="date-picker"
        type="button"
      >
         From Date 
      </button>
      <input
        class="ng-untouched ng-pristine ng-valid"
        container="body"
        formcontrolname="to"
        ngbdatepicker=""
        placement="bottom-right"
        type="hidden"
      />
      
      <button
        class="date-picker"
        type="button"
      >
         To Date 
      </button>
      <button
        type="submit"
      >
          
      </button>
    </form>
    
    
  </div>,
}
`;

exports[`FiltersComponent When rendering Then it matches snapshot 1`] = `
DebugElement {
  "nativeNode": <div
    id="root1"
  >
    <form
      class="ng-untouched ng-pristine ng-valid"
      novalidate=""
    >
      <select
        class="form-select ng-untouched ng-pristine ng-valid"
        formcontrolname="projectId"
      >
        <option
          value=""
        >
            
        </option>
        
      </select>
      <select
        class="form-select ng-untouched ng-pristine ng-valid"
        formcontrolname="gatewayId"
      >
        <option
          value=""
        >
            
        </option>
        
      </select>
      <input
        class="ng-untouched ng-pristine ng-valid"
        container="body"
        formcontrolname="from"
        ngbdatepicker=""
        placement="bottom-right"
        type="hidden"
      />
      
      <button
        class="date-picker"
        type="button"
      >
         From Date 
      </button>
      <input
        class="ng-untouched ng-pristine ng-valid"
        container="body"
        formcontrolname="to"
        ngbdatepicker=""
        placement="bottom-right"
        type="hidden"
      />
      
      <button
        class="date-picker"
        type="button"
      >
         To Date 
      </button>
      <button
        type="submit"
      >
          
      </button>
    </form>
    
    
  </div>,
}
`;
